test/functional/test_framework/ellsq.py

```diff
@@ -0,0 +1,289 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+# Source: https://github.com/sipa/writeups/tree/main/elligator-square-for-bn
+"""Test-only Elligator Squared implementation
+
+WARNING: This code is slow and uses bad randomness.
+Do not use for anything but tests."""
+
+import hashlib
+import os
+import random
+import unittest
+
+from .key import fe, ECPubKey, SECP256K1, SECP256K1_G, SECP256K1_ORDER
+
+C1 = fe(-3).sqrt()
+C2 = (C1 - fe(1)) / fe(2)
+B = fe(7)
+
+def forward_map(u):
+    """Forward mapping function
+
+    Parameters:
+        u (of type fe) : any field element
+    Returns:
+        fe, fe : affine X and Y coordinates of a point on the secp256k1 curve
+    """
+    s = u**2
+    x1 = C2 - C1*s / (fe(1)+B+s)
+    g1 = x1**3 + B
+    if g1.is_square():
+        x, g = x1, g1
+    else:
+        x2 = -x1 - fe(1)
+        g2 = x2**3 + B
+        if g2.is_square():
+            x, g = x2, g2
+        else:
+            x3 = fe(1) - (fe(1)+B+s)**2 / (fe(3)*s)
+            g3 = x3**3 + B
+            x, g = x3, g3
+    y = g.sqrt()
+    if y.is_odd() == u.is_odd():
+        return x, y
+    else:
+        return x, -y
+
+def reverse_map(x, y, i):
+    """Reverse mapping function
+
+    Parameters:
+        fe, fe : X and Y coordinates of a point on the secp256k1 curve
+        i      : integer in range [0,3]
+    Returns:
+        u (of type fe) : such that forward_map(u) = (x,y), or None.
+
+        - There can be up to 4 such inverses, and i selects which formula to use.
+        - Each i can independently from other i values return a value or None.
+        - All non-None values returned across all 4 i values are guaranteed to be distinct.
+        - Together they will cover all inverses of (x,y) under forward_map.
+    """
+    if i == 0 or i == 1:
+        z = fe(2)*x + fe(1)
+        t1 = C1 - z
+        t2 = C1 + z
+        if not (t1*t2).is_square():
+            return None
+        if i == 0:
+            if t2 == fe(0):
+                return None
+            if t1 == fe(0) and y.is_odd():
+                return None
+            u = ((fe(1)+B)*t1/t2).sqrt()
+        else:
+            x1 = -x-fe(1)
+            if (x1**3 + B).is_square():
+                return None
+            u = ((fe(1)+B)*t2/t1).sqrt()
+    else:
+        z = fe(2) - fe(4)*B - fe(6)*x
+        if not (z**2 - fe(16)*(B+fe(1))**2).is_square():
+            return None
+        if i == 2:
+            s = (z + (z**2 - fe(16)*(B+fe(1))**2).sqrt()) / fe(4)
+        else:
+            if z**2 == fe(16)*(B+fe(1))**2:
+                return None
+            s = (z - (z**2 - fe(16)*(B+fe(1))**2).sqrt()) / fe(4)
```

Comment:
I think this might be easier to read if ```fe(16)*(B+fe(1))**2)``` and ```z**2 - fe(16)*(B+fe(1))**2``` were assigned to variables.

Reply:
thanks @KevinMusgrave, that's true. it's built on top of #26222 now and much better readability.